kube-runtime/src/shared_store/prism.rs

```diff
@@ -0,0 +1,61 @@
+use crate::shared_store::ready_token::ReadyState;
+use crate::watcher;
+use crate::watcher::Event;
+use futures::stream::BoxStream;
+use futures::{stream, Stream, StreamExt};
+use std::fmt::Debug;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio::sync::broadcast;
+
+/// Wraps a Reflector stream and enables it to split out so that additional subscribers can be added.
+/// This allows Controllers to subscribe to the same stream of events as the Reflector, without
+/// having to own the original stream.
+///
+/// We only forward Ok events along the subscriptions and let the primary Reflector stream handle any
+/// errors.
+// todo - not sure about struct name?
+pub struct Prism<K> {
+    reflector: BoxStream<'static, watcher::Result<Event<K>>>,
+    ready_state: ReadyState,
+    sender: broadcast::Sender<Event<K>>,
+}
```

Comment:
I'll start here. This  Prism/SharedStore pattern can work, but it feels a bit inside out.

Ideally, this `Prism` logic can be a method on `WatchStreamExt`, rather than a public top level type because then it is compatible with both `reflector` and `watcher`. That might actually be pretty close to what you have with `WatchStreamExt::subscribe` returning a `Prism` AFAIKT since you `impl Stream` for it.

For the second problem; blocking until ready; I feel like that can maybe be done in `WatchStreamExt` as well. We already have `WatchStreamExt::backoff` there as a way to limit flow, so it should be possible to do something similar to wait until we have seen ONE element (a Once thing inside a struct that impls stream can maybe embed a token) without having the user manage/pass ready tokens around. That way we could conceivably do something like:

```rust
let cm_stream = reflector(writer, watcher(...)).once_ready().await;
let cm_stream_cpy = cm_stream.subscribe();
```

Reply:
(it would be a minor performance hit to do multiple `.once_ready().await`s sequentially, but I think this can be worked around with a `join_all` before creating subscriptions.)

Reply:
Cheers for the suggestions; I'll look at moving some of this stuff around onto the WatchStreamExt. I'm still on a pretty steep learning curve with streams so you'll have to bear with me on lots of boxed streams until I fully get my head wrapped around things properly 😅.

I have a concern with the `once_ready` suggestion in that it could be easy to forget, or not discovered, until you get bitten by a bug caused by a race condition.

If we're still thinking of something like the `SharedStore` struct that provides a wrapper over `Store` creation then that alleviate that concern somewhat. Easy mode would be `shared_store.store(lp).await` and if more fine gained control is wanted then it's perhaps not unreasonable to expect people to invest a bit more in learning the API.

The token approach does allow the Store creation to remain synchronous rather than async (although I'm not sure how valuable that actually is).

Reply:
Some more thoughts on `once_ready` (because it's not far from how we currently handle our stores and trying to implement something like it reminded me of some of the issues we had to tackle).

It would probably need to not only be async, but also return be resulty since the stream could end before the await completes. Our implementation also takes a timeout so that the future won't wait indefinitely if the stream never successfully receives an event, which is another type of failures that needs to be handled.

One of my aims with the initial approach was to defer these problems away from setup to the point where you start consuming the streams.

Reply:
Yeah, good point about the "missed `once_ready`" type mistake. Not sure what the best way for that would be. Possible you could do a wrapper type for the stream to encapsulate the "post-init" state, but that's pretty hairy to do on streams (maybe even too hairy to be worth it).

As for the `Result` problem, you could still collect futures (of `once_ready` results) early without awaiting them, then do a `try_join!` on all the streams before subscribing, but I guess that does not plug into `timeout` easily? We might have to supply a helper that does a raw `select!` to do this nicely possibly. It sounds brittle 😞

Reply:
I'm just going to re-iterate the specific problem we're trying to solve so it's clear in my head. 

When you call `.state()` on a `Store` you may get a result that hasn't yet being initialised. We're currently trying to solve the init problem (either by not running controllers until all the stores have reported ready, or by supplying a future that can be waited until a store is ready).

Perhaps we're tackling the problem in the wrong place? We could provide a wrapper over store (`SafeStore`?), or (more radically) change the interface of store itself so that calling `.state()` or `.get()` are async. They won't return a result until the `Store` has initialised but once it has will always instantly poll ready with the answer.

Reply:
I've had a think about the approach I started here and realised the I was bundling everything together specifically to solve the race condition problem. If you can solve the race condition in a different way you can unbundle everything.

Given that; I've pared things right back and deleted all of the Controller stuff for now to focus on the duplicate stores and race condition (which are separate concerns which probably belong in their own PRs, keeping it all here for now for historical reasons while it's discussed). 

I've added in a struct called `SafeStore` which exposes async versions of the `Store` methods. They will block until the `Store` has received the first event and then always instantly return the value. Instances of `SafeStore` can be given to controllers with no worry about race conditions.

I've also added a `subscribable` method in `watch_ext` which returns a version of the watch stream that exposes a `subscribe_ok` method. This will allow controllers to hook into the same stream that a store uses (probably with an approach something like you discussed in Issue 1080).

As I mentioned above `SharedStore` has dropped all notions of handling controllers. It now just provides `Stores`, with the same caching mechanism as before. Additional I've added in `subscribe` methods to the `SharedStore`, allowing controllers to pull their streams from them.

To revisit the original example I put above:

```rust
fn setup_pvc_controller(
    namespace: &str,
    pvc_shared_store: &mut SharedStore<PersistentVolumeClaim>,
    pod_shared_store: &mut SharedStore<Pod>,
    cm_shared_store: &mut SharedStore<ConfigMap>,
) {
    let pod_store = pod_shared_store.all(ListParams::default().labels(“app=nginx”));
    let cm_store = cm_shared_store.all(ListParams::default().labels(“app=other”));

    // Given that to setup a controller you'll always call both of these method, perhaps there should be a single
    // call that returns both the store and the subscription?
    let pvc_store = pvc_shared_store.namespaced(namespace, ListParams::default().labels(“app=nginx”))
    let pvc_stream = pvc_shared_store.namespaced_subscribe(namespace, ListParams::default().labels(“app=nginx”))

   // Below syntax is entire made up
   Controller::new_from(pvc_store, pvc_stream).run(....)
}
```